File: features/DungeonCommands.js
import Settings from "../config";
import { RegisterGroup } from "../utils/RegisterStuff";

const COOLDOWN_MS = 30000;

const Floor = {
  1: "ONE",
  2: "TWO",
  3: "THREE",
  4: "FOUR",
  5: "FIVE",
  6: "SIX",
  7: "SEVEN"
};

const InstanceType = {
  m: "MASTER_CATACOMBS",
  f: "CATACOMBS"
};

let cooldownEnd = 0;
let addedToQueue = false;

const dungeonCommands = new RegisterGroup({
  command: register("command", (arg) => {
    if (!arg) return ChatLib.command("warp dungeons");

    const match = /^([mf])([1-7])$/i.exec(arg);
    if (!match) return;

    const [_, prefix, num] = match;
    const floorNum = parseInt(num);
    const floorWord = Floor[floorNum];
    const instanceType = InstanceType[prefix.toLowerCase()];
    if (!floorWord || !instanceType) return;

    const now = Date.now();
    const timeLeft = Math.max(0, cooldownEnd - now);

    if (timeLeft > 0) {
      if (addedToQueue) return ChatLib.chat(`Rejoining in ${(timeLeft / 1000).toFixed(2)}s`);
      addedToQueue = true;
      ChatLib.command(`pc Waiting for cooldown, rejoining in ${(timeLeft / 1000).toFixed(2)}s.`);
    }

    setTimeout(() => {
      ChatLib.command(`joininstance ${instanceType}_FLOOR_${floorWord}`);
      enterUndersized();
    }, timeLeft + 100);
  }).setName("d").unregister(),

  chat: register("chat", (rank, name, dungeontype, number) => {
    if (name === Player.getName()) cooldownEnd = Date.now() + COOLDOWN_MS;
  }).setCriteria("${*}[${rank}] ${name} entered ${dungeontype} Catacombs, Floor ${number}${*}").unregister(),
});

Settings.registerListener("Dungeon Commands", v => v ? dungeonCommands.register() : dungeonCommands.unregister());
if (Settings.toggleDcmd) dungeonCommands.register();

function enterUndersized() {
  if (!Settings.toggleUndersizedEntry) {
    addedToQueue = false;
    return;
  }

  Client.scheduleTask(10, () => {
    Player.getContainer()?.click(13, false, "LEFT");
    addedToQueue = false;
  });
}

File: src/main/java/dev/slhj/slhjaddons/features/DungeonCommands.java
package dev.slhj.slhjaddons.features;

import com.mojang.brigadier.arguments.StringArgumentType;
import dev.slhj.slhjaddons.core.Feature;
import dev.slhj.slhjaddons.core.Setting;
import dev.slhj.slhjaddons.util.ClientUtils;
import dev.slhj.slhjaddons.util.McUtils;
import net.fabricmc.fabric.api.client.command.v2.ClientCommandRegistrationCallback;
import net.fabricmc.fabric.api.client.command.v2.ClientCommands;
import net.minecraft.client.player.LocalPlayer;
import net.minecraft.world.inventory.AbstractContainerMenu;
import net.minecraft.world.inventory.ContainerInput;

import java.util.regex.Pattern;

public final class DungeonCommands extends Feature {

    private static final Pattern FLOOR_PATTERN = Pattern.compile("^([mf])([1-7])$", Pattern.CASE_INSENSITIVE);
    private static final long COOLDOWN_MS = 30_000;

    private static final String[] FLOOR_NAMES = {"", "ONE", "TWO", "THREE", "FOUR", "FIVE", "SIX", "SEVEN"};

    private final Setting.ToggleSetting enterUndersizedSetting;

    private long cooldownEnd = 0;
    private boolean addedToQueue = false;

    public DungeonCommands() {
        setLabel("Dungeon Commands");
        category(Category.DUNGEONS);
        enterUndersizedSetting = toggle("dungeon_commands.enter_undersized", "Auto-Enter Undersized", false);
    }

    public static final String id = "dungeon_commands";
    @Override public String id() { return id; }

    @Override
    public void init() {
        ClientCommandRegistrationCallback.EVENT.register((dispatcher, access) ->
                dispatcher.register(ClientCommands.literal("d")
                        .then(ClientCommands.literal("help")
                                .executes(ctx -> { help(); return 1; }))
                        .executes(ctx -> { handleDungeonCommand(""); return 1; })
                        .then(ClientCommands.argument("floor", StringArgumentType.word())
                                .executes(ctx -> {
                                    handleDungeonCommand(StringArgumentType.getString(ctx, "floor"));
                                    return 1;
                                }))));
    }

    private void help() {
        McUtils.chat("&7Usage: /d [f1-f7 | m1-m7]");
        McUtils.chat("&7  /d f7");
    }

    public void handleDungeonCommand(String arg) {
        if (arg == null || arg.isEmpty()) {
            McUtils.runCommand("warp dungeons");
            return;
        }

        var matcher = FLOOR_PATTERN.matcher(arg);
        if (!matcher.matches()) {
            McUtils.chat("&cInvalid floor: " + arg + ". /dungeon help for usage.");
            return;
        }

        String prefix = matcher.group(1).toLowerCase();
        int floorNum = Integer.parseInt(matcher.group(2));
        String instanceType = prefix.equals("m") ? "MASTER_CATACOMBS" : "CATACOMBS";

        long now = System.currentTimeMillis();
        long timeLeft = Math.max(0, cooldownEnd - now);

        if (timeLeft > 0) {
            if (addedToQueue) {
                McUtils.chat(String.format("Rejoining in %.2fs", timeLeft / 1000.0));
                return;
            }
            addedToQueue = true;
            McUtils.runCommand(String.format("pc Waiting for cooldown, rejoining in %.2fs.", timeLeft / 1000.0));
        }

        long delay = timeLeft + 100;
        McUtils.scheduleTask(() -> {
            McUtils.runCommand(String.format("joininstance %s_FLOOR_%s", instanceType, FLOOR_NAMES[floorNum]));
            enterUndersized();
        }, delay);

        cooldownEnd = System.currentTimeMillis() + COOLDOWN_MS;
    }

    private void enterUndersized() {
        addedToQueue = false;
        if (!enterUndersizedSetting.value().get()) return;

        LocalPlayer player = ClientUtils.player();
        if (player == null) return;
        AbstractContainerMenu menu = player.containerMenu;
        McUtils.MC.gameMode.handleContainerInput(menu.containerId, 13, 0, ContainerInput.PICKUP, player);
    }
}